tests/boards/native_posix/cpu_wait/src/main.c

```diff
@@ -0,0 +1,295 @@
+	/* @ 150ms */
+	k_busy_wait(WASTED_TIME);
+}
+
+/**
+ * @brief Test the POSIX arch k_busy_wait and cpu_hold while another thread
+ * takes time during this test thread waits
+ *
+ * Note: This test relies on the exact timing of the ticks.
+ * For native_posix it works, with a tick of 10ms. In general this test will
+ * probably give problems if the tick time is not a relatively even number
+ * of microseconds
+ */
+static void test_cpu_hold_with_another_thread(void)
+{
+	uint64_t time2, time1;
+
+	k_sleep(Z_TIMEOUT_TICKS(1)); /* Wait until tick boundary */
+
+	/* Let TIME_WASTER thread start */
+	k_sem_give(&start_sema);
+
+	/* k_busy_wait part: */
+
+	time1 = posix_get_hw_cycle();
+	/* @40 ms */
+	k_busy_wait(2*ONE_TICK_TIME + 1);
+	/* The thread should have switched in and have used
+	 * WASTED_TIME us right after 2*one_tick_time
+	 * As that is longer than 1 tick + 1us, the total should be
+	 * 2 ticks + WASTE_TIME
+	 */
+
+	time2 = posix_get_hw_cycle();
+
+	zassert_true(time2 - time1 == 2*ONE_TICK_TIME + WASTED_TIME,
+			"k_busy_wait failed %llu-%llu != %u\n",
+			time2, time1, 2*ONE_TICK_TIME);
+
+	/* @61 ms */
+	k_sleep(Z_TIMEOUT_TICKS(1)); /* Wait almost 2 ticks */
+
+	time1 = posix_get_hw_cycle();
+	/* @80 ms */
+	k_busy_wait(ONE_AND_HALF_TICKS);
+	/* The thread should have used WASTED_TIME us after 1 tick,
+	 * but as that is lower than 1.5 ticks, in total the wait should be
+	 * 1.5 ticks
+	 */
+
+	time2 = posix_get_hw_cycle();
+
+	zassert_true(time2 - time1 == ONE_AND_HALF_TICKS,
+			"k_busy_wait failed %llu-%llu != %u\n",
+			time2, time1, ONE_TICK_TIME);
+
+	/* CPU hold part: */
+
+	/* @95 ms */
+	k_sleep(Z_TIMEOUT_TICKS(1)); /* Wait almost 2 ticks */
+
+	time1 = posix_get_hw_cycle();
+	/* @110 ms */
+	posix_cpu_hold(ONE_TICK_TIME + 1);
+	/* The thread should have used WASTED_TIME us after 1 tick,
+	 * so the total should be 1 tick + WASTED_TIME + 1 us.
+	 * That is compared to k_busy_wait() we spend 1 us more in this context
+	 */
+
+	time2 = posix_get_hw_cycle();
+
+	zassert_true(time2 - time1 == ONE_TICK_TIME + 1 + WASTED_TIME,
+			"k_busy_wait failed %llu-%llu != %u\n",
+			time2, time1, ONE_TICK_TIME);
+
+
+	/* @121.001 ms */
+	k_sleep(Z_TIMEOUT_TICKS(1)); /* Wait almost 2 ticks */
+
+	time1 = posix_get_hw_cycle();
+	/* @140 ms */
+	posix_cpu_hold(ONE_AND_HALF_TICKS);
+	/* The thread should have used WASTED_TIME us after 1 tick,
+	 * so the total wait should be 1.5 ticks + WASTED_TIME.
+	 * That is 1.5 ticks in this context, and WASTED_TIME in the other
+	 * thread context
+	 */
+
+	time2 = posix_get_hw_cycle();
+
+	zassert_true(time2 - time1 == ONE_AND_HALF_TICKS + WASTED_TIME,
+			"k_busy_wait failed %llu-%llu != %u\n",
+			time2, time1, ONE_TICK_TIME);
+}
+
+/**
+ * Replacement system tick timer interrupt handler which wastes time
+ * before calling the real one
+ */
+void np_timer_isr_test_replacement(const void *arg)
+{
+	ARG_UNUSED(arg);
+
+	k_busy_wait(WASTED_TIME);
+
+	void np_timer_isr_test_hook(const void *arg);
```

Comment:
Please move this function prototype somewhere out of this function.

Reply:
This is really a matter of stile :). Being a test hook/hack I prefer to have it as local as possible.
There is no good excuse to place it anywhere else. And if I place it in this file global scope, somebody may ask to place it in the driver header.

Reply:
It's just highly unusual/surprising to have a prototype inside a function. If you move the prototype just above line 194, it is still very local, but avoiding the element of surprise. But I agree that this is a matter of style :-)

Reply:
(You never expect the spanish inquisition :p)
Yes it is quite unusual :), but note that it actually changes the scope of the prototype, nobody can use it outside of here without a warning. So when reading the code you also know this is the only user of this prototype.

Reply:
Not done :) (I did place it on purpose there, to indicate to a future reader that it is only used just in the line below)

Reply:
Ok.